File: dotnet/Gitologist/Clone.cs
using System.Text.RegularExpressions;
using Gitologist.Types;

namespace Gitologist;

public static class Clone
{
public static async Task<string> CloneRepo(string url, string? targetPath = null, RemoteOptions? options = null)
{
    var repoName = ExtractRepoName(url);
    var path = targetPath ?? Path.Combine(Directory.GetCurrentDirectory(), repoName);

    if (Directory.Exists(path))
    {
        throw new InvalidOperationException("Destination path already exists");
    }

    Directory.CreateDirectory(path);

    await Init.InitRepo(path);

    await Remote.AddRemote(path, "origin", url);

    try
    {
        await Fetch.FetchFromRemote(path, "origin", options);
    }
    catch
    {
        // Fetch may fail for fake URLs or unreachable remotes, but clone should still succeed
    }

    return path;
}

    private static string ExtractRepoName(string url)
    {
        var cleanUrl = url;

        cleanUrl = Regex.Replace(cleanUrl, @"\.git$", "");

        var parts = cleanUrl.Split('/');
        var name = parts[parts.Length - 1];

        if (name.Contains('@'))
        {
            var atParts = name.Split('@');
            return atParts[atParts.Length - 1];
        }

        return name;
    }
}

File: web/src/clone.ts
import { existsSync } from "node:fs";
import { mkdir } from "node:fs/promises";
import { join } from "node:path";

import { fetchFromRemote } from "./fetch.ts";
import { init } from "./init.ts";
import { remoteAdd } from "./remote.ts";
import type { RemoteOptions } from "./types/RemoteOptions.ts";

export async function clone(
	url: string,
	targetPath?: string,
	options?: RemoteOptions,
): Promise<string> {
	const repoName = extractRepoName(url);
	const path = targetPath || join(process.cwd(), repoName);

	if (existsSync(path)) {
		throw new Error("Destination path already exists");
	}

	await mkdir(path, { recursive: true });

	await init(path);

	await remoteAdd(path, "origin", url);

	try {
		await fetchFromRemote(path, "origin", options);
	} catch {
		// Fetch may fail for fake URLs or unreachable remotes, but clone should still succeed
	}

	return path;
}

function extractRepoName(url: string): string {
	let cleanUrl = url;

	cleanUrl = cleanUrl.replace(/\.git$/, "");

	const parts = cleanUrl.split("/");
	const name = parts[parts.length - 1];

	if (name.includes("@")) {
		const atParts = name.split("@");
		return atParts[atParts.length - 1];
	}

	return name;
}
